src/grpc/stream_infer_handler.cc

```diff
@@ -140,6 +140,17 @@ ModelStreamInferHandler::Process(InferHandler::State* state, bool rpc_ok)
   // This means that we only need to take care of
   // synchronizing this thread and the ResponseComplete
   // threads.
+  // We need an explicit finish indicator. Can't use 'state->step_'
+  // because we launch an async thread that could update 'state's
+  // step_ to be FINISH before this thread exits this function.
+  bool finished = false;
+  if (state->context_->grpc_strict_) {
+    std::lock_guard<std::recursive_mutex> lock(state->context_->mu_);
+    // Check if stream error detected and already connection ended
+    if (state->context_->IsGRPCStrictError()) {
```

Comment:
Does this require both if statements? Or does `IsGRPCStrictError()` returning `true` require by `grpc_strict_` being true?

Reply:
grpc_strict_ is a bool per stream set by the user when they begin the stream, indicating that they want to use this feature. Once set it is true throughout the stream.
IsGRPCStrictError() is a function that checks if the particular stream has encountered an error from the core previously. We need to maintain this to avoid double close of the stream.
Hence need locking around it.